man/man1/prpath.1
.\"// $Id$
.\"// Copyright (c) 2001,2002                        RIPE NCC
.\"//
.\"// All Rights Reserved
.\"//
.\"// Permission to use, copy, modify, and distribute this software and its
.\"// documentation for any purpose and without fee is hereby granted,
.\"// provided that the above copyright notice appear in all copies and that
.\"// both that copyright notice and this permission notice appear in
.\"// supporting documentation, and that the name of the author not be
.\"// used in advertising or publicity pertaining to distribution of the
.\"// software without specific, written prior permission.
.\"//
.\"// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\"// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
.\"// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
.\"// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\"// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\"// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"//
.\"// 
.\"//  Copyright (c) 1994 by the University of Southern California
.\"//  and/or the International Business Machines Corporation.
.\"//  All rights reserved.
.\"//
.\"//  Permission to use, copy, modify, and distribute this software and
.\"//  its documentation in source and binary forms for lawful
.\"//  non-commercial purposes and without fee is hereby granted, provided
.\"//  that the above copyright notice appear in all copies and that both
.\"//  the copyright notice and this permission notice appear in supporting
.\"//  documentation, and that any documentation, advertising materials,
.\"//  and other materials related to such distribution and use acknowledge
.\"//  that the software was developed by the University of Southern
.\"//  California, Information Sciences Institute and/or the International
.\"//  Business Machines Corporation.  The name of the USC or IBM may not
.\"//  be used to endorse or promote products derived from this software
.\"//  without specific prior written permission.
.\"//
.\"//  NEITHER THE UNIVERSITY OF SOUTHERN CALIFORNIA NOR INTERNATIONAL
.\"//  BUSINESS MACHINES CORPORATION MAKES ANY REPRESENTATIONS ABOUT
.\"//  THE SUITABILITY OF THIS SOFTWARE FOR ANY PURPOSE.  THIS SOFTWARE IS
.\"//  PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES,
.\"//  INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\"//  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, TITLE, AND 
.\"//  NON-INFRINGEMENT.
.\"//
.\"//  IN NO EVENT SHALL USC, IBM, OR ANY OTHER CONTRIBUTOR BE LIABLE FOR ANY
.\"//  SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES, WHETHER IN CONTRACT,
.\"//  TORT, OR OTHER FORM OF ACTION, ARISING OUT OF OR IN CONNECTION WITH,
.\"//  THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"//
.\"//  Questions concerning this software should be directed to 
.\"//  info-ra@isi.edu.
.\"//
.\"//  Author(s): Cengiz Alaettinoglu <cengiz@isi.edu>
.\"
.\"
.TH prpath 1 local
.SH NAME
prpath
.SH SYNOPSIS
.B prpath
.RI [ options ]\ [[<as-no>]\ <destination-specification>]
.SH DESCRIPTION
.PP
prpath enumerates a list of paths between <as-no>
and <destination-specification>.
<destination-specification> can be an AS number,
an IP address or an address prefix.
If <as-no> is not specified it is deduced for your
system from the IP address of your host and IRR. If no
<destination-specification> is given prpath goes into a command loop where
it inputs <destination-specification> from stdin, one at a line.
.PP
prpath ignores error and warning messages due to communication to the
database server or parsing policy objects.
Please see \-ignore_errors and \-report_errors options below.
.SH IRRToolSet COMMON OPTIONS
.RS
.IP -help
Print summary of command-line options and abort.
.IP "\-T [whois_query | whois_response | input | all]"
Trace the argument. Useful for debugging.
.IP "\-D <integer>"
Print debugging messages of debug channel <integer> 
(if compiled with DEBUG defined).
.IP "\-version"
Print the version number and quit.
.IP "\-h <host-name>"
Establish an radbserver connection to host <host-name>.
The default is whois.radb.net.
.IP "\-p <port-no>"
Establish an radbserver connection to port <port-no>.
The default is 43.
.IP "\-protocol <protocol>"
Use the <protocol> to connect to the IRR server. <protocol> can be irrd,
ripe, and bird. The default is irrd. The ripe option should be used with
the ripe server.
.IP "\-s <source-list>"
Consider the sources specified in the comma separated <source-list>.
If an object is defined in multiple sources in <source-list>,
prpath uses the definition first encountered in <source-list>
from left to right.
.IP \-rusage
Print resource usage upon completion.
.IP "\-prompt <string>"
Set the interactive prompt to <string>.
.IP "\-ignore_errors"
Do not print error and warning messages due to communication to the
database server or parsing policy objects.
.IP "\-report_errors"
Print error and warning messages due to communication to the
database server or parsing policy objects.
.RE
.SH OTHER OPTIONS
.RS
.IP "\-as <as-no>"
Use AS <as-no> as a source AS.
.IP "\-mpl <integer>"
If the shortest path between the source and the destination is N hops,
report all paths that are shorter than N + <integer>.
.RE
.SH ENVIRONMENT VARIABLES
.RS
.IP IRR_HOST
Specifies the radbserver host to connect.
.IP IRR_PORT
Specifies the radbserver port number to connect.
.IP IRR_SOURCES
Specifies the source list (comma separated) to consider.
.RE
.PP
Command line options take precedence over environment variables.
.SH EXAMPLES
.PP
.nf
	% ./prpath AS2914 193.85.128.0/24
	Route:
	   NLRI: 193.85.128.0/24
	   Origins: AS2819 
	   Communities: COMM_NSFNET 
	AS2819 AS286 AS2914
.fi
.PP
.SH ERROR AND WARNING MESSAGES
Warning: No autnum record found for AS
.RS
.PP
The AS specified in nonexistent in the specified Database for specified soure list
.PP
.RE
Unable to lookup destination AS
.RS
.PP
If you do not specify which AS is destination AS, prpath tries to query IRR to get originating autnum. If route is non-existent, you get this error.
If you specify <ipv4_prefix> as <destination>, prpath will try to find origin of this prefix. If you specify <ipv4_address> as <destination>, prpath will try to find origin of prefix <ipv4_address>/32.
.PP
.RE
Unknown protocol!
.PP
.RS
Connection to IRR server failed. This can be caused by various reasons, see Errors manpage, IRR Communication errors.
.RE
.PP
For more error descriptions, please see
.B
Errors manual page.
.SH AUTHORS
Cengiz Alaettinoglu <cengiz@isi.edu>
.PP
Katie Petrusha <katie@ripe.net>
.SH ACKNOWLEDGEMENTS
This is enhanced version of RIPE's prpath tool.


NEWS
Release Notes for IRRToolSet
****************************

5.1.0 - 2016-04?-NN
------------------
* Starting with this release, IRRToolSet is going to follow a Semantic
  Versioning scheme: http://semver.org/
* Initial support for IOS XR config generation, use -config ciscoxr
* Rename @rtconfig set commands and add JunOS Support:
  cisco_prefix_acl_no to prefix_acl_no
  cisco_aspath_acl_no to aspath_acl_no
  cisco_pktfilter_acl_no to pktfilter_acl_no
  cisco_community_acl_no to cisco_community_acl_no
* Cisco IOS improvements:
  Move from peer-groups to peer templates
  Add support for community.delete()
  Reduced default starting id's for community lists, acl's etc from 100 to 50
  Added support for BGP communities starting with 0:
  Fix bug where both IPv4 & IPv6 prefixes would be added to the same prefix-list
* JunOS improvements:
  -junos_replace to output replace: configuration that can be loaded with "load replace"
  -junos_and_not_or to make route policies that AND community and prefix-lists to match
   the rtconfig Cisco behaviour.
  Increment JunOS policy-statement id's for IPv4/v6 to avoid duplicate id's
  Add support for IPv6 prefix-length-range to bring it to parity with IPv4
  Fixed minor configuration output ordering
  Fix problems with 4-byte AS's

5.0.1
-----
* don't barf on pingable: and ping-hdl: attributes (RFC 5943)
* don't barf on inet6num: assignment-size (RIPE policy proposal 2010-06)
* RFC 5952 requires that ipv6 addresses be presented in lower case.
* ipv6 addresses are now printed in optimised format.
* added geoloc: and language: fields to both inetnum: and inet6num: objects.
* compilation fixes for g++ 4.7 / debian wheeze
* stop talking asdot to IRRDBs.  none of the IRRs supports asdot since ~2010
* silence a bunch of warnings for clang 3.0
* fix AS output for rpslcheck

5.0.0
-----
Substantial reorganisation of package.  Only rtconfig, peval and rpslcheck survived. 
For further information on changes for 5.0.0, please consult the irrtoolset
wiki on http://irrtoolset.isc.org/

Moreover, IRRToolSet was switched to a new MIT style license, from a
non-commercial license previously. Thanks to past author Cengiz Alaettinoglu
and partial copyright holder, University of Southern California, for graciously
arranging this.

Miscellanous fixes:
* fix f_junos.cc to make export and import token names consistent. Thanks to
  Stan Barber.
* Support for 4 byte ASNs. Many thanks to S.P.Zeidler for this.
* Makefile now honors DESTDIR. Thanks to Kaj Niemi for the patch.
* @RtConfig fix for IPv6. Thanks to Nick Hilliard for the report and
  to Timo Koehler for the patch.
* Fix when an IOS as-path list from an RPSL as-set contains more
  than one AS number. Thanks to Hagen Boehm for the patch.
* Compile fix for Solaris 9 and other systems that do not have
  stdint.h defined. Thanks to Hagen Boehm for reporting the problem.
* Fix for core dump on Solaris. Thanks to Matthew Moyle-Croft for the
  report and to S.P.Zeidler for the fix.
* Fix for invalid RtConfig output on Cisco when specifying family.
  Thanks to Daryl Collins for the report and to S.P.Zeidler for the
  fix.
* Cache file support when "irrd" protocol is used. Thanks to Hagen
  Boehm for the implementation.
* Build fixes for OS X.
* Build fix for gcc 4.2 (default on Debian and Ubuntu these days).
* Explicit check for flex and bison in configure.

4.8.5
-----
Bug fix for RPSLng queries on IRRd servers.

Better support for 64-bit compilation.

Integrated a number of fixes from the NetBSD port as proposed by 
Havard Eidnes.

* Numerous compiler fixes:
  o src/config.h.in is amended to conditionally define __P(x).  Use
    __P() instead of _G_ARGS() in various places.
  o Many, many files are patched to replace occurrances of _G_uint32_t
    with uint32_t. The former is in the host compiler implementation's
    name space, and as far as I know, a library writer has no business
    assuming anything about the types defined in that name space
    (names starting with a single underscore character).
  o Instead of including <_G_config.h>, use <config.h>, and add
    includes of e.g. <sys/types.h> where needed.
  o A few troublesome casts removed which newer C++ compilers balk at.

* JunOS support enabled & fixed

* Portability to Dragonfly BSD (treat as FreeBSD)

* Bugfix for Cisco export

4.8.4
-----
* A number of runtime problems which caused crashes were reported, and
  these have been fixed. Thanks to Hagen Boehm for discovering them.

4.8.3
-----
A set of patches collected by Hagen Boehm was applied:

* whois.apnic.net server problem
  APNIC's whois server (RIPE NCC software with RPSS functions
  disabled) has changed syntax of some embedded comments in a way that
  violates the protocol assumptions in IRRToolSet 4.8.2
  (required/allowed "empty lines" ~ "\n\n" in the context of "%"
  comment lines. This problem seems to be around since APNIC switched
  to a database version supporting IPv6/RPSLng).
  Thanks to Hagen Boehm.

* ASnnn embedded in AS-Set AS-xxxx doesn't resolve properly (prefix
  loss)
  We are seeing cases where resolving to route-sets (a) an ASnnn and
  (b) an AS-set AS-xxxx which includes ASnnn as a member lists some
  prefixes as members of ASnnn but NOT as members of AS-xxxx (Which
  turns something like "125.16.0.0/14^14-24" into some thousand
  separately listed 125.16.x.y/n^n-n or 125.16.x.y/n^m-m that don't
  combine into any 125.16.x.y/n^m-l unless m=l).
  Thanks to Hagen Boehm.

* multi protocol filter problem 
  prefix filter consisting of ipv4 and ipv6 prefixes get not resolved
  properly.  Either only prefixes of a single protocol get resolved
  (e.g. only ipv4) or the route-set contains ANY. The first case was
  observed when resolving an "ASxxxx" filter, the latter case when
  resolving an "ASxxxx ASyyyy" filter, where both ASxxxx and ASyyyy
  contained ipv4 and ipv6 prefixes.
  Thanks to Hagen Boehm.

* due to an uninitialized integer the irrd client showed erroneous

branches/spz-work/src/prpath/prpath.cc
//  $Id$
// Copyright (c) 2001,2002                        RIPE NCC
//
// All Rights Reserved
//
// Permission to use, copy, modify, and distribute this software and its
// documentation for any purpose and without fee is hereby granted,
// provided that the above copyright notice appear in all copies and that
// both that copyright notice and this permission notice appear in
// supporting documentation, and that the name of the author not be
// used in advertising or publicity pertaining to distribution of the
// software without specific, written prior permission.
//
// THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
// ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
// AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
// DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
// AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
//
//
//  Copyright (c) 1994 by the University of Southern California
//  All rights reserved.
//
//  Permission to use, copy, modify, and distribute this software and its
//  documentation in source and binary forms for lawful non-commercial
//  purposes and without fee is hereby granted, provided that the above
//  copyright notice appear in all copies and that both the copyright
//  notice and this permission notice appear in supporting documentation,
//  and that any documentation, advertising materials, and other materials
//  related to such distribution and use acknowledge that the software was
//  developed by the University of Southern California, Information
//  Sciences Institute. The name of the USC may not be used to endorse or
//  promote products derived from this software without specific prior
//  written permission.
//
//  THE UNIVERSITY OF SOUTHERN CALIFORNIA DOES NOT MAKE ANY
//  REPRESENTATIONS ABOUT THE SUITABILITY OF THIS SOFTWARE FOR ANY
//  PURPOSE.  THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
//  IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE,
//  TITLE, AND NON-INFRINGEMENT.
//
//  IN NO EVENT SHALL USC, OR ANY OTHER CONTRIBUTOR BE LIABLE FOR ANY
//  SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES, WHETHER IN CONTRACT, TORT,
//  OR OTHER FORM OF ACTION, ARISING OUT OF OR IN CONNECTION WITH, THE USE
//  OR PERFORMANCE OF THIS SOFTWARE.
//
//  Questions concerning this software should be directed to 
//  ratoolset@isi.edu.
//
//  Author(s): Cengiz Alaettinoglu <cengiz@ISI.EDU>
//             John Mehringer <mehringe@isi.edu>

#include "config.h"

#include <typeinfo>
#include <iostream>
#include <iomanip.h>
#include <cstring>
#include <cctype>


#include <cstdlib>
extern "C" {
#if HAVE_UNISTD_H
#   include <unistd.h>
#endif
}

#include "irrutil/Argv.hh"
#include "irrutil/net.hh"
#include "util/rusage.hh"
#include "irrutil/debug.hh"
#include "irrutil/trace.hh"
#include "irrutil/version.hh"

#include "rpsl/schema.hh"

#include "irr/irr.hh"
#include "irr/autnum.hh"
#include "irr/rawhoisc.hh"
#include "irr/ripewhoisc.hh"

#include "bgp/bgproute.hh"

//IRR * irr; // Global variable defined in irr/irr.cc
             // Used for bug 
            
static ASt source_as;  // AS number of source for prpath trace
static IRR * whois;    // Connection to IRR server for query
static ActionDictionary * actiondictionary;
Rusage ru;

static char * destination_arg;

static IPAddr * destination_ip = NULL;
static Prefix * destination_prefix = NULL;
static ASt destination_as;

int  opt_rusage        = 0;
char *opt_prompt       = "prpath> ";
char *opt_my_as        = NULL;
int  opt_echo_as       = 0;
int  opt_max_path_length = 8;
int  path_length_limit = 12;
int  opt_interactive = 1;

extern int BGPROUTE_MAX_ASPATH_LENGTH;

void Announce(BGPRoute &bgproute, ASt as);


//--------------------------------------------------------------
// void usage(char *argv[])
//
//
//--------------------------------------------------------------
void usage(char *argv[]) {
   cerr << "Usage: " << argv[0] << " to be written later" << endl;
   exit(1);
}

//--------------------------------------------------------------
// int start_tracing(char *dst, char *key, char *nextArg)
//
//
//--------------------------------------------------------------

int start_tracing(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      trace.enable(nextArg);
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0; 
}

//--------------------------------------------------------------
// int start_debugging(char *dst, char *key, char *nextArg)
//
//
//--------------------------------------------------------------

int start_debugging(char *dst, char *key, char *nextArg) {
   if (nextArg) {
      Debug(dbg.enable(atoi(nextArg)));
      return 1; // return 1 to signify nextArg is used by us
   }
   return 0;
}


//--------------------------------------------------------------
// void init_and_set_options (int argc, char **argv, char **envp)
//
//
//--------------------------------------------------------------

void init_and_set_options (int argc, char **argv, char **envp) {
   ArgvInfo argTable[] = {
      // RAToolSet common arguments
      // key, type, src, dst, help
      {"-T", ARGV_FUNC, (char *) &start_tracing,      (char *) NULL, 
       "Start tracing the next argument"},
      {"-D", ARGV_FUNC, (char *) &start_debugging,    (char *) NULL, 
       "Start debugging the next argument"},
      {"-version", ARGV_FUNC, (char *) &version,      (char *) NULL,
       "Show version"},
      {"-rusage", ARGV_BOOL, (char *) NULL,           (char *) &opt_rusage,
       "On termination print resource usage"},
      {"-prompt", ARGV_STRING,  (char *) NULL,        (char *) &opt_prompt,
       "Prompt"},

      IRR_COMMAND_LINE_OPTIONS,

      // prpath specific arguments
      {"-as",  ARGV_STRING,       (char *) NULL, (char *) &opt_my_as,
       "AS number of the aut-num object to use."},
      {"-mpl", ARGV_INT,          (char *) NULL, (char *) &opt_max_path_length,
       "Max path length"},

      {(char *) NULL, ARGV_END, (char *) NULL, (char *) NULL, (char *) NULL}
   };

   IRR::handleEnvironmentVariables(envp);
   
   IRR::IgnoreErrors(NULL, NULL, NULL);

   if (ParseArgv(&argc, argv, argTable, 0) != ARGV_OK) {
      cerr << endl;
      exit(1);
   }
  
   // Setup Connection to IRR server
   if (! (whois = IRR::newClient())) {
      cerr << "Unknown protocol!" << endl;
      exit(1);
   }

   if (opt_max_path_length) 
     BGPROUTE_MAX_ASPATH_LENGTH = opt_max_path_length;

   // if there are remaining arguments
   // the first one is my as number
   // the second one is a destination prefix
   switch (argc) {
     case 3 :
       opt_my_as = argv[1];
     case 2:
       destination_arg = argv[argc - 1];
       if ((destination_arg[0] == 'A' || destination_arg[0] == 'a') && 
           (destination_arg[1] == 'S' || destination_arg[1] == 's')) {
         destination_as = atoi(destination_arg + 2);
       } else if (strchr(destination_arg,'/')) {
         destination_prefix = new Prefix(destination_arg);
       } else {
         destination_ip = new IPAddr(destination_arg);
         destination_prefix = new Prefix(destination_ip->Prefix::get_text());
       }
       opt_interactive = 0;
     case 1:
       break;
     default:
       cerr << "Wrong number of arguments..." << endl;
       exit(-1);
   }
   
   if (opt_my_as)
     source_as = atoi(opt_my_as + 2);
   else {
     ipAddr myip;
     myip.setToMyHost();
     if ((source_as = whois->getOrigin("%s/32", myip.getIpaddr())) == 0) {
	    cerr << "prpath could not guess your AS number, "
	            "please specify an AS number on the command line." 
	         << endl;
	    delete whois;
	    exit(1);
	  }
   }      

   // have a prompt only if the input is coming from a tty
   if (!opt_interactive || !isatty(fileno(stdin)) || !isatty(fileno(stdout)))
      opt_prompt = NULL;
}

//-------------------------------------------------------------
// void Import(BGPRoute &bgproute, ASt export_as, ASt import_as,
//             IPAddr *import_ip, IPAddr *export_ip)
//
//  export_as was the current as import_as is the peer_as 
//  from IterateExports and 
//
//  Now we have to check to see if the peer as will import the
//  current as since BGP works from destination to source
//
//  import_ip & export_ip are the ip addresses of the routers
//      
//     also look at CIDRAdvisor/proxyagg.cc for similar code
//-------------------------------------------------------------
void Import(BGPRoute &bgproute, ASt export_as, ASt import_as,
            const IPAddr *import_ip, const IPAddr *export_ip) {
  const AutNum * import_autnum = NULL;
  // Check to see if loop is created by adding this AS 
  if (! bgproute.loopCheck(import_as)) {  
    // Get Imports from IRR 
    import_autnum = whois->getAutNum(import_as);
    if (import_autnum) {
      AutNumSelector<AttrImport> import_itr(import_autnum, "import", NULL,
					    export_as, export_ip, import_ip);
      if (const FilterAction * fa = import_itr.first()) {
              for (; fa; fa = import_itr.next()) {
        // if not at max nodes in route 
        // then continue depth first search
          if (!bgproute.atMaxRouteLength()) {
            if (bgproute.isMatching(import_as, fa->filter, whois)) {
              if (import_as == bgproute.source) { // Have we found the source
                 // If we have found the source as then print out our route      
                bgproute.print();
              } else {
                bgproute.executeAction(fa->action, actiondictionary);
                Announce(bgproute, import_as);
              }
            } // end if (bgproute.isMatching())
          } else {
            //cerr << "  Max Path length reached." << endl;
          }
        } // end for
        
      } else {  //No Imports then Check defaults
      // cerr << "   No Imports! Checking for default..." << endl;
        AttrIterator<AttrDefault> itr(import_autnum, "default");
        for (const AttrDefault *  ad = itr.first(); ad; ad = itr.next()) {
          if (! bgproute.atMaxRouteLength()) {
            if (bgproute.isMatching(import_as,ad->filter, whois)) {
              if (import_as == bgproute.source) { // Check to see if we found the source
                bgproute.print();
              } else {
                bgproute.executeAction(ad->action, actiondictionary);
                Announce(bgproute, import_as);
              }
            } // end if (bgproute.isMatching
          } else {
            //cerr << "Max Path length reached." << endl;
          }
        }  // end for
      }
    }  // end if (import_autnum)
  }  // end if (! bgproute.loopCheck())
}

//--------------------------------------------------------------
//  void IterateExports(BGPRoute &bgproute, ASt as, AutNum * as_autnum,
//                      ASt peer_as, IPAddr * peer_ip, IPAddr * local_ip)
//
//    Find export filter which contain peer_as,
//    Loop over list of exports,
//    Execute the actions for that export filter
//
//    peer_ip & local_ip are ip addresses for the routers
//-------------------------------------------------------------
void IterateExports(BGPRoute &bgproute, ASt as, const AutNum * as_autnum,
                    ASt peer_as, const IPAddr * peer_ip, 
		    const IPAddr * local_ip) {
  // Get matching Export Filter-Actions
  AutNumSelector<AttrExport> export_itr(as_autnum, "export", NULL, peer_as, peer_ip, local_ip);
  const FilterAction * fa;
  BGPRoute * import_route;
  // Check each Export filter for possible correct route
  for (fa = export_itr.first(); fa; fa = export_itr.next()) {
    import_route = new BGPRoute(bgproute); // Duplicate the currrent BGPRoute 
                                           // to check this possible path

    // For each export check if they allow the current AS's  
    // traffic on their network by looking at the peer AS's imports
    // Import(bgproute, current as, ip of current as, peer AS, peer IP address)
    import_route->executeAction(fa->action, actiondictionary);
    Import(*import_route, as, peer_as, peer_ip, local_ip);
    delete import_route;  // We are done with it
  }  
}  

//--------------------------------------------------------------
//  void Announce(BGPRoute &r, ASt as) 
//    Analogous to a bgp AS announcement
//      Check exports of current AS to see where to go next
//      doing a recursive depth first search on each export  
//      it terminates when either the max route length is reached
//      or the BGP's source AS is found
//
//   Announce()-->IterateExports()-->Imports()-->Announce()...until path is found 
//-------------------------------------------------------------
void Announce(BGPRoute &bgproute, ASt as) {
   const AutNum * as_autnum = whois->getAutNum(as);
   ASt peer_as;
   bgproute.add(as);  // Add the current AS as to the BGPRoute path list

   if (as_autnum) {  // Do we have an autnum record for this as?
      // Find all peers of current AS
      AutNumPeeringIterator itr(as_autnum);
      for (const Peering *peering = itr.first(); itr; peering = itr.next()){
	 // Get exports containing peerAS and check the imports of the peer
	 IterateExports(bgproute, as, as_autnum,
			peering->peerAS, &peering->peerIP, &peering->localIP);
      } // End for
   } else 
      cerr << "Warning: No autnum record found for AS" << as << endl;
}
  

//-------------------------------------------------------------
// main (int argc, char **argv, char **envp)
//
//
//-------------------------------------------------------------
main (int argc, char **argv, char **envp) {
  const AutNum * myAutNum;  
  BGPRoute * bgproute;

  schema.initialize();
    
  init_and_set_options(argc, argv, envp);
  
  irr = whois;  // For irr/autnum.hh  AutNumIterator::isMatching(Attribute *attr)

  actiondictionary = new ActionDictionary;
 
  if (destination_prefix != NULL) {
    bgproute = new BGPRoute(source_as, destination_prefix, whois);
  } else {
    bgproute = new BGPRoute(source_as, destination_as);
  }

 /* myAutNum = irr->getAutNum(2764);
  AutNumPeeringIterator itr(myAutNum);
  for (const Peering *peering = itr.first(); itr; peering = itr.next()) {
     cout << "AS" << peering->peerAS;
     if (!peering->peerIP.isNull())
	cout << " " << peering->peerIP;
     if (!peering->localIP.isNull()) 
	cout << " at " << peering->localIP;
     cout << "\n";
  }
   
  exit(1);*/

  // by katie@ripe.net - if route is non-existent
  if (bgproute->origin->list.head() == NULL)
	cout << "Unable to lookup destination AS" << endl;

  // Start the search from destination to the source AS
  for (ItemASNO * asln = bgproute->origin->list.head(); asln;
                  asln = bgproute->origin->list.next(asln)) {
    if (asln->asno) {
      cout << "Route:" << endl;
      cout << "  NLRI: ";
      
      if (destination_prefix)
        cout << destination_arg << endl;
      else
        cout << "don't care" << endl;
        
      cout << "  Origins: AS" << asln->asno << endl; 

      Announce(*bgproute, asln->asno); 
    } else
      cout << "Unable to lookup destination AS" << endl;
  }
  
  delete irr;
  delete actiondictionary;
}
